Improve losc support in search workflow

ahnitz opened the pull request:
Things that will be done:
     * Allow LOSC to be used as a segment database alternative
     * segments and veto information are taken in at runtime by direct call instead of external program execution. This significantly reduces the code complexity in some areas. 
     * veto definer file not required for LOSC useage, optional for dqsegdb
          * veto definer file will define "macros" of flags to make interface uniform (TODO)
     * unify api between LOSC and segdb (TODO)

ahnitz commented:
*Note*: This will require some configuration file changes. I'll post those and include a PR against pycbc-config when this PR is closer to a feature complete state.

spxiwh commented:
Hey @ahnitz. I agree with the general proposal outlined in this PR. But I have one comment at this point:

The proposed format for the `[workflow-segments]` section is confusing. It is not obvious why `DATA:CBC_CAT1_VETO` would mean `DATA - CBC_CAT1_VETO`. Is it possible to give some more thought into how to make this more obvious what is being done?

ahnitz commented:
@spxiwh Yes, absolutely. We could actually just allow "DATA - CBC_CAT1_VETO" type syntax. Would that work?

ahnitz commented:
@spxiwh I've changed the syntax to the following. Hopefully this meets with your approval now. 

```
science-segments = '+DATA1,+DATA2,-VETO1,-HWINJ,-VETO2'. 
```

This was easier to parse and I think still clear.

ahnitz commented:
@duncan-brown @spxiwh Ok, I've tested the LOSC side of this. An example data.ini file now looks like. 

```
[workflow-segments]
segments-science = +DATA,-CBC_CAT1_VETO
segments-vetoes = +CBC_CAT2_VETO,+CBC_HW_INJ,+BURST_HW_INJ
```

For dqsegdb it should be the same, except the names need to come from dqsegdb or the veto definer file. It uses whatever is in the name field of the definer file as a possibility. One can also give the standard veto definer and server options to specify those for dqsegdb, though the server does now default to segments.ligo.org if not given. One could in principle do an analysis without a veto definer and just give the dqsegdb flags, though that would currently omit the possibility of adding extra padding options. 

If you just use LOSC flags it does not require any server, veto definer, or dqsegdb installed.

ahnitz commented:
@spxiwh Also note that this complete removes the use of segments_from_cats and segment_query executables.

ahnitz commented:
Test config is stored here. https://www.atlas.aei.uni-hannover.de/work/ahnitz/apdf/new/. 

I should note that the flag names for dqsegdb must also include the version number, but not the ifo i.e.

```
SOME_FLAG:1,SOME_OTHER_FLAG:2
```

spxiwh commented:
Codeclimate is reporting a number of issues that do need fixing (bad indentation, redefining variables, whitespace, ...). Probably this needs rebasing(?)

spxiwh commented on pycbc/events/status.py:
```diff
@@ -0,0 +1,252 @@
+# =============================================================================
+#
+""" Utitlities to query the status of gravitational-wave instruments
+from public sources as well as dqsegdb.
```
This gives me a misleading impression of what this does. I read this and think that I can query (in realtime) if instruments are operational or not. Most of the stuff here is a little more technical (parsing veto-definers/querying segments). Also think the name "status" is confusing/misleading.

... It also doesn't make sense to have this in the "events" package. This is segments querying so we should use our name for this ("segments" ... or even "flags", but have this in a "segments" module, or a top level "segments.py")

spxiwh commented on pycbc/workflow/segment.py:
```diff
@@ -1389,3 +1390,58 @@ def file_needs_generating(file_path, cp, tags=None):
+         veto_definer = save_veto_definer(workflow.cp, out_dir, [])
+
+    # Check for provided server
+    server = "segments.ligo.org"
```
Should this not default to "https://segments.ligo.org, as it used to?

ahnitz replied:
I can't test this, but I think it should be without the url header given that the query command itself has a separate input for this part.

spxiwh commented on pycbc/workflow/segment.py:
```diff
@@ -1389,3 +1390,58 @@ def file_needs_generating(file_path, cp, tags=None):
+    
+    segments = {}
+    for ifo in workflow.ifos:
+        flag_str = cp.get_opt_tags("workflow-segments", option_name, [ifo])
```
So if H1 and L1 (for e.g.) are using different definitions for science (V1 certainly will for internal usage) we would need to have `[workflow-segments-h1]`, `[workflow-segments-l1]` and `[workflow-segments-v1]`? As this is a top-level `workflow` section, there's no reason to not allow the current `segments-science-h1` syntax, but either way it should be clear (e.g. make the example assume that H1/L1/V1 definitions are different even if they are not).

spxiwh replied:
.... I also deliberately kept the functionality that you can use H1 flags (and data) and tell the workflow it is V1 (for example). I think you've removed that ... It might be useful.

spxiwh replied:
(maybe I'm saying keep the old syntax and enforce a full segment string e.g. `H1:SCIENCE:1`)

ahnitz replied:
I think redundancy should be avoided where possible, but I'm not against allowing an easy override. I propose to allow the following to make this self-consistent.

H1:SCIENCE:1 will just use this flag for whatever detector you had said. 
SCIENCE:1 standard form, will use the ifo that is was requested a flag for

spxiwh commented on pycbc/events/status.py:
```diff
@@ -0,0 +1,252 @@
+        segs = query_flag(ifo, flag_name, start_time, end_time,                                       
+                          source=source, server=server,
+                          veto_definer=veto_definer)
+        if total_segs is not None:
```
Why not just initialize `total_segs` as an empty segmentlist?

spxiwh commented on pycbc/events/status.py:
```diff
@@ -0,0 +1,252 @@
+                                      veto_definer=veto_definer)
+    down_flags = [] if down_flags is None else down_flags
+    for flag_name in down_flags:
+        mseg = query_flag(ifo, flag_name, start_time, end_time)
```
I'm unclear why this doesn't also use `query_cumulative_flags`.

.... This seems to allow for there to be some uncertainties here. For example if I do:

```science=SCIENCE-CAT_1+MISLABELED_CAT_1_I_ACTUALLY_WANT_TO_ANALYSE```

I think I would not get `MISLABELED_CAT_1_I_ACTUALLY_WANT_TO_ANALYSE`. In other words if a segment is added *and* subtracted in multiple places the behaviour seems ill defined.

ahnitz replied:
Yes, it is ill defined then. I've added to the documentation that positive time-adding flags must precede negations.

spxiwh commented on pycbc/events/status.py:
```diff
@@ -0,0 +1,252 @@
+    segments: glue.segments.segmentlist
+        List of segments
+    """
+    if source == 'GWOSC' or source == 'any':
```
Source=`any` is not defined, but is the default?? What is this?

spxiwh replied:
Indeed it seems to mean `GWOSC`, which seems wrong.

ahnitz replied:
If you give 'any' it tries GWOSC first to find the flag, then proceeds to dqsegdb if that fails.

spxiwh commented on pycbc/events/status.py:
```diff
@@ -0,0 +1,252 @@
+            from dqsegdb.apicalls import dqsegdbQueryTimes as query
+        except ImportError:
+            raise ValueError("Could not query flag. DQSegdb is not installed" 
+                             "to additionally try there")
```
Make this clearer. "You need to install dqsegdb!"

spxiwh commented on pycbc/events/status.py:
```diff
@@ -0,0 +1,252 @@
+        # We treat the veto definer name as if it were its own flag and
+        # a process the flags in the veto definer
+        flag_segments = segmentlist([])
+        if veto_definer is not None and segment_name in veto_def[ifo]:
```
If I'm right you're going to be querying the entire veto-definer file multiple times. This is a slow step and should only be run once.

spxiwh commented on pycbc/events/status.py:
```diff
@@ -0,0 +1,252 @@
+            data[ifo[i]] = {}
+        
+        if category[i] not in data[ifo[i]]:
+            data[ifo[i]][category[i]] = []
```
So category is just an integer so data['H1']['1'] would contain all the CAT_1 vetoes. So IIRC the macro name for `CAT_1` is currently `1`. Think this needs a `"CAT_"+` or similar here

spxiwh replied:
(And then there's the Hardware injections are still `3` problem. In the old code we remap that to "H")

spxiwh replied:
Okay, I have to make some change here because the macro name for CAT_1 is currently `int(1)` which cannot be specified as a string!

ahnitz replied:
@spxiwh How hard is that to change? It would be nice if we could get rid of this discrepency already

spxiwh replied:
Not something that's in the scope of this merge request. We can ask about this, but it requires coordination with everyone else using these, so this would have to support the current version now.

spxiwh commented:
Would also need to edit the new workflow generator in coordination with Marton

spxiwh commented on pycbc/events/status.py:
```diff
@@ -0,0 +1,252 @@
+            except:
+                raise ValueError("Could not query flag, check name (%s) or times",
+                                 segment_name)
+
```
I ended up here having apparently not gone down any of these if branches ... Need an else: Raise Error here

ahnitz replied:
I can't match up your comment with the line. Do you mean you got this exception? Or do you mean something else?

spxiwh commented on pycbc/workflow/segment.py:
```diff
@@ -1389,3 +1390,58 @@ def file_needs_generating(file_path, cp, tags=None):
+        down = [x[1:] for x in flags if x[0] == '-']        
+
+        key = ifo + ':' + name
+        segments[key] = query_combined_flags(ifo, up, start, end, down,
```
down is a key-word argument. Please declare as such

spxiwh commented on pycbc/events/status.py:
```diff
@@ -0,0 +1,252 @@
+                                  source=='dqsegdb')
+            else:
+                raise ValueError(msg)
+    elif source == 'dqsegdb':
```
I don't think I can ever get here, as source is not set anywhere!

spxiwh commented on pycbc/events/status.py:
```diff
@@ -0,0 +1,252 @@
+            msg = "Unable to find segments in GWOSC, check flag name or times"
+            if source == 'any':
+                return query_flag(ifo, segment_name, start_time, end_time,
+                                  source=='dqsegdb')
```
Ah, I see, you'd fall through here ... But you dropped a bunch of important key-word arguments ... Somehow I've fallen through with source=False ... Oh I see, you used `==`, DOH!

... I'm editing some of this locally now.

ahnitz replied:
Uh... please don't edit locally that will conflict with me.

duncan-brown commented:
@ahnitz if the goal is to get rid of veto definer files completely and switch to this method of specifying segments, then we should check that this patch doesn’t regresss functionality.

With a veto definer file, you could:

 * Specify the IFO, name, and version of a flag. I assume this would just be rolled into the name here, but I don’t see an explicit syntax convention for version.
 * Specify the GPS interval over which to apply a DQ flag, even if it is defined over a broader interval. We’ve used this functionality in production searches in the past.
 * Specify a (possibly asymmetric) padding for a DQ flag (e.g. or extend hardware injections by +8 and -4 second.)

It’s proably worth @spxiwh looking at a veto definer file to see if there’s any metadata in there that I’ve forgotten about.

spxiwh commented:
@duncan-brown This does not remove a veto-definer, it's just optional now (although this is a little confusing). I am running now with a VDF, although I do have some complaints, which I'm logging.

Another one to add. The resulting "SCIENCE" segment file (which now has CAT_1 removed) has not stored what times are SCIENCE, and which are CAT_1 times. It is useful to see this breakdown of information (and allows the user to see how the SCIENCE file was produced).

spxiwh commented:
I also need to check that all the things Duncan notes are still being done ...

spxiwh commented:
@ahnitz Please hold off on fixes while I'm testing. Please revert e6226f9 for a minute, I'll open a pull request from here.

ahnitz commented:
@duncan-brown There is no such thing as an IFO or vesion number with LOSC flag names. Those can only be given for dqsegdb flag names.

ahnitz commented:
@spxiwh Let me know when I have the token back to implement changes.

spxiwh commented on pycbc/workflow/segment.py:
```diff
@@ -1389,3 +1390,58 @@ def file_needs_generating(file_path, cp, tags=None):
+                                             server=server,
+                                             veto_definer=veto_definer)
+        logging.info("%s: got %s flags", ifo, option_name) 
+
```
I got a nasty and unclear error if setting the veto categories to the wrong names ie:
```
vetoes=+BLAH,+BLOOPY
```

spxiwh replied:
I *think* this is also resolved, but needs double checking

spxiwh commented:
Adding a requirement: The information in Tables A3 and A4 on the summary pages must still be producable (although we can be more clear about what these various categories mean!)

spxiwh commented:
The segment summary information is not being stored. That should be added.

spxiwh commented:
The `hlevel plotting` stuff is taking ages to generate. It doesn't normally take this long and I'm running on vulcan, so no filesystem problems.

```
2018-11-20 11:57:01,605:INFO : ...leaving coincidence 
2018-11-20 11:57:43,657:INFO : hlevel plotting: 0
2018-11-20 12:02:01,458:INFO : hlevel plotting: 1
```

I don't see how this could be related to this patch, but I've not seen this before.

spxiwh commented:
The `VETOES` file needs a proper name:
```
H1L1-VETOES-0-1.xml
```
(start and end time need to be set correctly)

ahnitz commented:
@spxiwh How did you get a file named that? I can't reproduce that. Also hlevel taking forever, I have no idea about. I don't see how it's related to this patch.

spxiwh commented:
> @spxiwh How did you get a file named that? I can't reproduce that. Also hlevel taking forever, I have no idea about. I don't see how it's related to this patch.

I just ran the code as normal, basically. Maybe it's because the file is empty? (Like I said, there's still a problem here). The `hlevel` stuff is now taking almost 30 minutes!!

spxiwh commented on pycbc/workflow/segment.py:
```diff
@@ -1389,3 +1390,60 @@ def file_needs_generating(file_path, cp, tags=None):
+
+        flags = flag_str.replace(' ', '').strip().split(',')
+        up = [x[1:] for x in flags if x[0] == '+']
+        down = [x[1:] for x in flags if x[0] == '-']
```
You need to make sure that len(up) + len(down) == len(flags). I forgot the "+"s.

spxiwh commented:
The problem with the `hlevel` stuff seems to be a rather large number of tags being sent to these jobs. I guess that's unrelated, but not sure why it's causing a problem now and not previously??

ahnitz commented:
@spxiwh I've moved it and renamed to dq.py. I don't like segments as it's unclear what it is. It is information that says what the state of the instrument is. That sounds like dq to me. I'm open to other options.

spxiwh commented:
`dq` -> `data_quality`

ahnitz commented:
@spxiwh @duncan-brown The rest of the climate issues, I'm ignoring.

duncan-brown commented:
@ahnitz, LOSC *should* provide versioning, it’s one of my requests, so don’t code out the ability to have that. 

Also, even though LOSC treats CAT1 as per of science, the do separate between CAT1 and DATA so you could handle it symmetrically and give @spxiwh what he wants in the tables. 

Also, how do you handle the fact that LOSC provides the logical NOT of the LSC segment definition for DQ? Assume that in PyCBC and internally invert?

duncan-brown commented:
@spxiwh BTW, I don't know if you have LOSC ticket access, but I filed this a while ago
```
[LDG #11769] Is there any plan to fix the LOSC segment JSON to allow the use of standard tools for querying LOSC (and e.g. adding versioning)?
```
I was told by LOSC they didn't have the human resources to support external users. Next time I get this complaint, you might want to remind them of the effort that @ahnitz is going to in order to *not* break internal LSC functionality that is different from the LOSC code.

duncan-brown commented:
@ahnitz how about the following?
```
+H1:SOME_FLAG:1[-3,+3],+SOME_OTHER_FLAG:2[0,0]
```
I would allow the user to give an IFO name.

ahnitz commented:
@spxiwh @duncan-brown Can you please have another look at this. The following syntax now should work.
```
science-segments = +L1:data:2<-8:8>[0:1000000000]
```
<> for padding, [] for duration. These are optional. The version and ifo are also optional (ifo matched to detector if not provided). (version has no meaning for GWOSC flags in any case). You can use tags if you want to supply different tags to each ifo as normal.

spxiwh commented:
I guess the full syntax is messy, but I don't have a better suggestion. This does seem to do what was requested from the syntax (though some tests of this would be helpful, extra bonus points if it can go into the test suite).

ahnitz commented:
@spxiwh I'll work on some unittests for the functions and some documentation. As for the syntax, I think most of the time this won't be needed. As far as we know is that the remainder to get this patch merged?

spxiwh commented:
Codeclimate is reporting some genuine issues to fix.

I would also still like to see time lost to CAT_2 or CAT_1, but as agreed before these would be separate calls in plotting parts of the workflow and not part of the main flow.

ahnitz commented:
@spxiwh I've added some unittests which will not be included in travis for the query related functions and moved almost all the logic in there so the workflow side is very simple now. I've also tested by hand that I can generate  a workflow. 

I think the next thing is to add some documentation.

ahnitz commented:
@spxiwh Ok, I updated the documentation and also went ahead and change how the workflow testing was done so that it points to the 'test' folder in pycbc-config now. Hopefully, going forward it should be clear that these ini files need to be kept working with master. 

Possibly related to #2422 

I haven't added any new code to make the table you are requesting (i.e. to show time lost due to different flags). Is that something required for this PR?

spxiwh commented:
@ahnitz Are we ready to move forward with this now?

It would be good to update:
`pycbc_create_offline_search_workflow` as well if possible, as that should soon replace the old code.

codeclimate is still reporting issues, but the linked page doesn't seem correct, and it's clear that multiple of these issues are not ones to be addressed, so I think I'm happy here.

spxiwh commented:
And just to check. I know we talked about removing XML from segment stuff, but a bunch of the plotting codes still expect XML inputs. Is it true that while the segment queries might no longer be writing/using XML files we will still output XML files for the plotting codes?

(Changing that can come later)

ahnitz commented:
@spxiwh  Yes, I would like to merge this. I don't want to keep adding more to this patch, so I think a followup would be to add this to the multi ifo workflow. 

Also, yes this doesn't completely remove xml, just from the query process itself. One step at a time. The coinc code for example still needs to read the xml files, though likewise we could replace that down the line with an hdf or json file.

spxiwh commented:
Okay, I just want to understand might break after merging this.

What's your thoughts on the developmental O3 workflow generator?

ahnitz commented:
@spxiwh Nothing should break in the multi-ifo workflow generator, and this code should be straightforward to also implement there (since it's largely a replacement for the older code). I didn't remove the older code in this patch. It's also used by a couple other workflows I think.

spxiwh commented:
Okay, I'm now happy to merge this.

This patch has inputs from multiple people, so I need to use the rebase and merge button. But for that we need to clean up the patch history. Can you collapse this into 3 commits. Alex's patch -> Ian's patch -> Alex's changes, then I'll merge.

ahnitz commented:
@spxiwh Can I merge this?
